crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,744 @@
+use crate::utils::{DualIpAddr, MutableIpPacket};
+use pnet_packet::{
+    ip::IpNextHeaderProtocols, ipv4::MutableIpv4Packet, ipv6::MutableIpv6Packet,
+    udp::MutableUdpPacket,
+};
+use std::{
+    hash::{DefaultHasher, Hash, Hasher},
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    time::Duration,
+};
+use telio_utils::{telio_log_warn, LruCache};
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The packet is too short to be a valid IP packet
+    PacketTooShort,
+    /// The packet has an unexpected IP version
+    UnexpectedIpVersion,
+    /// The packet has an unexpected transport protocol
+    UnexpectedTransportProtocol,
+    /// Cannot map the packet
+    CannotMap,
+    /// There is no mapping for the given port
+    MappingNotFound,
+}
+
+pub trait Nat {
+    /// Translate incomming packet (from the transport socket to the multicast peer)
+    /// Change the source to the multicast peer's ip and natted port
+    fn translate_incoming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+    /// Translate outgoing packet (from the multicast peer to the transport socket)
+    /// Change the destination to the peer's original ip and port
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+}
+
+type Mapper = fn(src_addr: SocketAddr) -> u16;
+
+fn default_mapper(src_addr: SocketAddr) -> u16 {
+    let mut hasher = DefaultHasher::new();
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
```

Comment:
So we are full controll of all port's so it should be fine to use a full u16 range